Pdep spin conservation

donerancl opened the pull request:
### Motivation or Problem
For pressure dependent networks, strange reactions that do not conserve spin are created. 

Examples:
N2 + HNO(S) (+M) <=> N2 + HNO(T) (+M)
HNO(S)  + HNO(S) (+M) <=> N2 + HNO(T) + HNO(T) (+M)
HNO(S) + N2H2 (+M) <=> HNO(T) + N2H2 (+M)
HNO(S)  + HNO(S) (+M) <=> N2 + HNO(S) + HNO(T) (+M)
...

This happens when intermediates can be formed from reactants with different multiplicities.

For example,
HNO(S) + NH2 <=> N2H3O
HNO(T) + NH2 <=> N2H3O
...
HNO(S) + NH2 <=> HNO(T) + NH2


### Description of Changes
In `arkane/explorer.py`, in the function `execute`, in the same loop that checks for forbidden structures,  I calculate the number of unpaired electrons on each side of the reaction. If there are >1 reactants and >1 products, and the number of unpaired electrons changes from reactants to products, I add the reaction to the list to be removed. 

Edit: I think this part of the code is only checking `path_reactions` which are directly connected. The problematic reactions are `net_reactions` 

### Testing
I tested this by generated a mechanism with pressure dependence and then running a pressure dependence job with Arkane.

In the resulting chem.inp file, one of the problematic reactions is commented out, but its reverse is not

```
! HNO(T)(118)+NH2(8)(+M)<=>HNO(15)+NH2(8)(+M)
HNO(15)+NH2(8)(+M)<=>HNO(T)(118)+NH2(8)(+M)
```

### Reviewer Tips
Suggestions for verifying that this PR works or other notes for the reviewer.

<!--
Checklist before submission:
 - Have you added appropriate unit tests?
 - Have you checked that all unit tests pass?
 - Is your code commented and understandable?
 - Have you updated related documentation?
 - Are the commits logically organized and informative?
 - Is your branch up to date with main?
-->

donerancl commented:
@xiaoruiDong Hi Xiaorui! I was wondering if you had any insights on this? It seems like my branch can remove (comment out) reactions that don't pass the check, but the reverse reaction is still there.

donerancl commented:
I updated the PR to check for `net_reactions` in `update` of `rmgpy/rmg/pdep.py`. This way, problematic reactions are not included under "Net reactions" in the log file, but they are still in the chemkin files...


From RMG.log file:
```PDepNetwork #311 network information
------------------------------------
Isomers:
    [NH]ON[O](301)                                        101.923 kJ/mol
    H2NONO(100)                                          -123.082 kJ/mol
Reactant channels:
    HNO(15) + HNO(15)                                    -13.0081 kJ/mol
    HNO(15) + HNO(T)(118)                                 67.5738 kJ/mol
    HNO(T)(118) + HNO(T)(118)                             148.156 kJ/mol
    NH(7) + HNO2(23)                                      88.9139 kJ/mol
    NO2(21) + NH2(8)                                     -7.20362 kJ/mol
    NO(14) + H2NO(17)                                    -71.8271 kJ/mol
Product channels:
    NH(7) + [O]N[O](227)                                  403.022 kJ/mol
    O(28) + [NH]O[NH](294)                                 438.48 kJ/mol
    O1NON1(450)                                           94.3616 kJ/mol
    H(5) + [NH]ON=O(298)                                  216.674 kJ/mol
    [NH]O[N]O(299)                                        73.3657 kJ/mol
    NO[N][O](307)                                         51.3043 kJ/mol
    [O-][NH+]1NO1(451)                                     76.242 kJ/mol
Path reactions:
    [NH]ON[O](301) <=> HNO(15) + HNO(15)                  154.012 kJ/mol
    NH(7) + [O]N[O](227) <=> [NH]ON[O](301)               403.022 kJ/mol
    O(28) + [NH]O[NH](294) <=> [NH]ON[O](301)              438.48 kJ/mol
    [NH]ON[O](301) <=> O1NON1(450)                        109.835 kJ/mol
    [NH]ON[O](301) <=> H2NONO(100)                        165.323 kJ/mol
    H(5) + [NH]ON=O(298) <=> [NH]ON[O](301)               268.974 kJ/mol
    HNO(15) + HNO(T)(118) <=> [NH]ON[O](301)              101.942 kJ/mol
    HNO(T)(118) + HNO(T)(118) <=> [NH]ON[O](301)          170.025 kJ/mol
    [NH]ON[O](301) <=> [NH]O[N]O(299)                     215.895 kJ/mol
    [NH]ON[O](301) <=> NO[N][O](307)                      266.705 kJ/mol
    [NH]ON[O](301) <=> [O-][NH+]1NO1(451)                 103.711 kJ/mol
    NH(7) + HNO2(23) <=> [NH]ON[O](301)                   118.797 kJ/mol
    [NH]O[N]O(299) <=> H2NONO(100)                        98.3389 kJ/mol
    NO2(21) + NH2(8) <=> H2NONO(100)                     -7.20363 kJ/mol
    NO(14) + H2NO(17) <=> H2NONO(100)                    -71.8271 kJ/mol
    H(5) + [NH]ON=O(298) <=> H2NONO(100)                  216.674 kJ/mol
Net reactions:
    HNO(15) + HNO(15) => [NH]ON[O](301)
    HNO(15) + HNO(15) => O1NON1(450)
    HNO(15) + HNO(15) => H2NONO(100)
    HNO(15) + HNO(15) => [NH]O[N]O(299)
    HNO(15) + HNO(15) => NO[N][O](307)
    HNO(15) + HNO(15) => [O-][NH+]1NO1(451)
========================================================================
```

from chemkin input: 
```
HNO(15)+HNO(15)(+M)<=>HNO(15)+HNO(T)(118)(+M)       1.000e+00 0.000     0.000    
HNO(15)+HNO(15)(+M)<=>HNO(T)(118)+HNO(T)(118)(+M)   1.000e+00 0.000     0.000   
```

xiaoruiDong commented:
> @xiaoruiDong Hi Xiaorui! I was wondering if you had any insights on this? It seems like my branch can remove (comment out) reactions that don't pass the check, but the reverse reaction is still there.

Hi Anna, I just saw this and found you added a few updates. Is this solved or help still needed?

hwpang commented:
My guess would be to check whether the filtered out reaction is still in either reaction_model.core.reactions or reaction_model.edge.reactions, and that causes that reaction to be added again somewhere that is not filtered

donerancl commented:
As an update, some concerning reactions should be allowed. For example, 
HNO(T) + NH2(D) <=> HNONH2 (D) <=> HNO(S) + HNO(D) 
is allowed if the unpaired electrons in the triplet and doublet have opposite spin. However, we often only have literature rate coefficients for one side and not the other. 

Reactions such as
HNO(T) + HNO(T) <=> [NH]OO[NH] <=> HNO(S) + HNO(T) 
are clearly incorrect though. The issue comes from the misidentification of the intermediate formed from combining two triplets of opposite spin as a triplet (which could produce HNO(S) + HNO(T)) instead of a singlet diradical which could produce either HNO(S) + HNO(S) or HNO(T) + HNO(T). I have begun to address this in a pull request draft on RMG-database. https://github.com/ReactionMechanismGenerator/RMG-database/pull/662

Another issue I have addressed in this pull request is that the pressure dependence code will treat the 1,2_Birad_alkene family as reversible. However, it is meant to represent irreversible intersystem crossings, like HNO(T) => HNO(S). I have added a check that sets pdep to False for irreversible reactions.

alongd commented:
Hi @donerancl,
I think we should update RMG somewhere deeper than Arkane/Explorer, perhaps in `generate_reactions()`.
I would create a dict (perhaps a YANL file) of allowed spin conversions (e.g. `S + S = [S]`, `D + D = [S, T]`, ...), and add a function that checks a reaction for spin conservation. This way we won't add "regular" reactions (if that was ever a problem, I don't know) nor unimolecular reactions that violate spin-conservation.

I'm not sure I know what to think about the example you showed:
```
HNO(S) + NH2(D) <=> N2H3O(D)
HNO(T) + NH2(D) <=> N2H3O(D)
...
HNO(S) + NH2(D) <=> HNO(T) + NH2(D)
```
Both path reactions are valid, but intuitively we don't like the net reaction which allegedly isn't conserving the spin. But if we give high credence to each path reaction (we definitely might not...) then the net reaction could be feasible. In other words, in addition to checking each path reaction for spin conservation, we should examine the mechanism and rate of each path reaction in these strange cases, and understand why RMG gives it a relatively high rate.
Happy to chat more.

donerancl commented:
Hi @alongd. Thanks for your comment. I think adding a spin check to arkane/explorer.py would be good. I will add that to this PR. 
I know what you mean about the net reactions. It seems clear that the S+D products would be more favorable unless producing species like NH, O, O2, which we know to be more stable as triplets. I think it's possible that poor rate estimates for the less favorable pathways, which often lack literature values, may cause fast rates for these strange net reactions. 

In RMG-database, I have a [draft PR](https://github.com/ReactionMechanismGenerator/RMG-database/pull/662) that will prevent some of these reactions from being generated in the first place. Particularly, the Birad_R_Recombination family currently allows triplet + triplet recombinations (e.g. O(T) + HNO(T) <=> [O]O[NH]). The product should be a singlet diradical since opposite spin is required to form a bond. However, when RMG updates the product molecule, it counts unpaired electrons + 1 as the multiplicity, making the reaction violate spin conservation (e.g. O(T) + HNO(T) <=> [O]O[NH](T)). The triplet radical can decompose giving O(T) and HNO(S), which gives the net reaction O(T) + HNO(T) <=> O(T) + HNO(S), which we don't want. I am working on incorporating a function to generate closed shell representations of singlet diradicals so that triplet and singlet PDep networks don't get scrambled.

Similarly, H_abstraction and Disproportionation families often generate pairs of reactions such as...
O2(T) + H2NO(D) <=> HO2(D) + HNO(T)                      H_abstraction
O2(T) + H2NO(D) <=> HO2(D) + HN=O(S)                   Disproportionation

[Cavallotti2023](https://www.sciencedirect.com/science/article/pii/S1540748922003078) calculated H-abstraction rates where they assumed that the reactants have opposite spin, so on the doublet surface, which would lead to disproportionation products. But [Sarathy2022](https://www.sciencedirect.com/science/article/pii/S001021802100451X). calculates an analogous rate on the quartet surface, giving H_abstraction products. In many similar cases (e.g. NH2 + H2NO <=> NH3 + HNO), we have literature rates for only one of these, usually the doublet surface. I think in most cases, it makes sense that the doublet surface pathway would be dominant. I worry that rates estimated for the quartet surface may be estimated poorly and compete with the probably more favorable doublet surface reaction. I suppose the options are to either improve the quartet surface rate estimate or to just assume only the doublet surface reaction is important.

alongd commented:
Thanks for these insights. Indeed very complex. Do you think it's worth telling RMG to differentiate between spin states when it trains the family tree and then makes estimations? We're short of data, but the some we have could be effective in this way. If so, perhaps we could think of representing the spin state somehow in the new RMG-db project?

ChgCao commented:
> Hi @alongd. Thanks for your comment. I think adding a spin check to arkane/explorer.py would be good. I will add that to this PR. I know what you mean about the net reactions. It seems clear that the S+D products would be more favorable unless producing species like NH, O, O2, which we know to be more stable as triplets. I think it's possible that poor rate estimates for the less favorable pathways, which often lack literature values, may cause fast rates for these strange net reactions.
> 
> In RMG-database, I have a [draft PR](https://github.com/ReactionMechanismGenerator/RMG-database/pull/662) that will prevent some of these reactions from being generated in the first place. Particularly, the Birad_R_Recombination family currently allows triplet + triplet recombinations (e.g. O(T) + HNO(T) <=> [O]O[NH]). The product should be a singlet diradical since opposite spin is required to form a bond. However, when RMG updates the product molecule, it counts unpaired electrons + 1 as the multiplicity, making the reaction violate spin conservation (e.g. O(T) + HNO(T) <=> [O]O[NH](T)). The triplet radical can decompose giving O(T) and HNO(S), which gives the net reaction O(T) + HNO(T) <=> O(T) + HNO(S), which we don't want. I am working on incorporating a function to generate closed shell representations of singlet diradicals so that triplet and singlet PDep networks don't get scrambled.
> 
> Similarly, H_abstraction and Disproportionation families often generate pairs of reactions such as... O2(T) + H2NO(D) <=> HO2(D) + HNO(T) H_abstraction O2(T) + H2NO(D) <=> HO2(D) + HN=O(S) Disproportionation
> 
> [Cavallotti2023](https://www.sciencedirect.com/science/article/pii/S1540748922003078) calculated H-abstraction rates where they assumed that the reactants have opposite spin, so on the doublet surface, which would lead to disproportionation products. But [Sarathy2022](https://www.sciencedirect.com/science/article/pii/S001021802100451X). calculates an analogous rate on the quartet surface, giving H_abstraction products. In many similar cases (e.g. NH2 + H2NO <=> NH3 + HNO), we have literature rates for only one of these, usually the doublet surface. I think in most cases, it makes sense that the doublet surface pathway would be dominant. I worry that rates estimated for the quartet surface may be estimated poorly and compete with the probably more favorable doublet surface reaction. I suppose the options are to either improve the quartet surface rate estimate or to just assume only the doublet surface reaction is important.



> Hi @alongd. Thanks for your comment. I think adding a spin check to arkane/explorer.py would be good. I will add that to this PR. I know what you mean about the net reactions. It seems clear that the S+D products would be more favorable unless producing species like NH, O, O2, which we know to be more stable as triplets. I think it's possible that poor rate estimates for the less favorable pathways, which often lack literature values, may cause fast rates for these strange net reactions.
> 
> In RMG-database, I have a [draft PR](https://github.com/ReactionMechanismGenerator/RMG-database/pull/662) that will prevent some of these reactions from being generated in the first place. Particularly, the Birad_R_Recombination family currently allows triplet + triplet recombinations (e.g. O(T) + HNO(T) <=> [O]O[NH]). The product should be a singlet diradical since opposite spin is required to form a bond. However, when RMG updates the product molecule, it counts unpaired electrons + 1 as the multiplicity, making the reaction violate spin conservation (e.g. O(T) + HNO(T) <=> [O]O[NH](T)). The triplet radical can decompose giving O(T) and HNO(S), which gives the net reaction O(T) + HNO(T) <=> O(T) + HNO(S), which we don't want. I am working on incorporating a function to generate closed shell representations of singlet diradicals so that triplet and singlet PDep networks don't get scrambled.
> 
> Similarly, H_abstraction and Disproportionation families often generate pairs of reactions such as... O2(T) + H2NO(D) <=> HO2(D) + HNO(T) H_abstraction O2(T) + H2NO(D) <=> HO2(D) + HN=O(S) Disproportionation
> 
> [Cavallotti2023](https://www.sciencedirect.com/science/article/pii/S1540748922003078) calculated H-abstraction rates where they assumed that the reactants have opposite spin, so on the doublet surface, which would lead to disproportionation products. But [Sarathy2022](https://www.sciencedirect.com/science/article/pii/S001021802100451X). calculates an analogous rate on the quartet surface, giving H_abstraction products. In many similar cases (e.g. NH2 + H2NO <=> NH3 + HNO), we have literature rates for only one of these, usually the doublet surface. I think in most cases, it makes sense that the doublet surface pathway would be dominant. I worry that rates estimated for the quartet surface may be estimated poorly and compete with the probably more favorable doublet surface reaction. I suppose the options are to either improve the quartet surface rate estimate or to just assume only the doublet surface reaction is important.

@donerancl Hi, Anna. Thanks for your detailed explanation! I am wondering whether the reaction `O(T) + HNO(T) <=> [O]O[NH]` you mentioned as an example of the triplet + triplet recombination → triplet diradical (not spin-conserving reaction) comes from the final RMG model, i.e. Chemkin file. I checked the recent generated RMG models and couldn't find out this reaction and the species [O]O[NH] in the dictionary (I think HNO2 in the dictionary does not refer to [O]O[NH]). Or this reaction may come from Edge reactions instead?

ChgCao commented:
> Hi @alongd. Thanks for your comment. I think adding a spin check to arkane/explorer.py would be good. I will add that to this PR. I know what you mean about the net reactions. It seems clear that the S+D products would be more favorable unless producing species like NH, O, O2, which we know to be more stable as triplets. I think it's possible that poor rate estimates for the less favorable pathways, which often lack literature values, may cause fast rates for these strange net reactions.
> 
> In RMG-database, I have a [draft PR](https://github.com/ReactionMechanismGenerator/RMG-database/pull/662) that will prevent some of these reactions from being generated in the first place. Particularly, the Birad_R_Recombination family currently allows triplet + triplet recombinations (e.g. O(T) + HNO(T) <=> [O]O[NH]). The product should be a singlet diradical since opposite spin is required to form a bond. However, when RMG updates the product molecule, it counts unpaired electrons + 1 as the multiplicity, making the reaction violate spin conservation (e.g. O(T) + HNO(T) <=> [O]O[NH](T)). The triplet radical can decompose giving O(T) and HNO(S), which gives the net reaction O(T) + HNO(T) <=> O(T) + HNO(S), which we don't want. I am working on incorporating a function to generate closed shell representations of singlet diradicals so that triplet and singlet PDep networks don't get scrambled.
> 
> Similarly, H_abstraction and Disproportionation families often generate pairs of reactions such as... O2(T) + H2NO(D) <=> HO2(D) + HNO(T) H_abstraction O2(T) + H2NO(D) <=> HO2(D) + HN=O(S) Disproportionation
> 
> [Cavallotti2023](https://www.sciencedirect.com/science/article/pii/S1540748922003078) calculated H-abstraction rates where they assumed that the reactants have opposite spin, so on the doublet surface, which would lead to disproportionation products. But [Sarathy2022](https://www.sciencedirect.com/science/article/pii/S001021802100451X). calculates an analogous rate on the quartet surface, giving H_abstraction products. In many similar cases (e.g. NH2 + H2NO <=> NH3 + HNO), we have literature rates for only one of these, usually the doublet surface. I think in most cases, it makes sense that the doublet surface pathway would be dominant. I worry that rates estimated for the quartet surface may be estimated poorly and compete with the probably more favorable doublet surface reaction. I suppose the options are to either improve the quartet surface rate estimate or to just assume only the doublet surface reaction is important.

@donerancl Hi, Anna. Thanks for your detailed explanation! I am wondering whether the reaction O(T) + HNO(T) <=> [O]O[NH] you mentioned as an example of the triplet + triplet recombination → triplet diradical (not spin-conserving reaction) comes from the final RMG model, i.e. Chemkin file. I checked the recent generated RMG models and couldn't find out this reaction and the species [O]O[NH] in the dictionary (I think HNO2 in the dictionary does not refer to [O]O[NH]). Or this reaction may come from Edge reactions instead?

alongd commented:
Hi @ChgCao,
A triplet + triplet giving a singlet product does (or at least may) conserve the spin if we have triplet(up, up) + triplet(down, down).
Our issue is that the edge has many cases like that and other cases which do violate spin conservation, and that they participate in our PDep networks. Then these PDep networks give core well-skipping reactions that are undesired (may or may not be spin-conserving)

donerancl commented:
Hi @alongd! I spoke with WHG yesterday, and I was not quite understanding the triplet-triplet spin conservation before. I found the following two papers that were very helpful. [Schulten 1983](https://www.ks.uiuc.edu/Publications/Papers/PDF/SCHU83/), [Bossanyi et al. JACS 2021](https://pubs.acs.org/doi/epdf/10.1021/jacsau.1c00322). Triplet triplet associations produce singlets, triplets, and quintets, in a 1:3:5 ratio _if_ strongly exchange coupled (the JACS paper does simulations for the weakly exchange coupled case, where the ratio can vary). "The 1/9 of random encounter pairs in the singlet state proceeds very fast to the annihilation products. The remaining fraction 8/9 of triplet and quintet pairs can annihilate only after change of the overall spin state." The quintet pairs are likely energetically infeasible like in the JACS example for rubrene and can be assumed to only dissociate back to the triplet pair. The triplet pairs can undergo an intersystem crossing to form a singlet pair that can dissociate, giving two singlets. For a quick glance, Figure 8 in the JACs paper and 1.20 in the Schulten paper are the most helpful.

I wonder what you think about introducing a spin statistics factor (potentially 1/9 + estimate for intersystem crossing) to the pre-exponential for triplet + triplet reactions such as O(T) + HNO(T) => HNOO(S)? I think in this case we would ignore O(T) + HNO(T) => HNOO(T), assuming some portion becomes HNOO(S) and the remaining goes back to O(T) + HNO(T). But in a pressure dependent network I wonder if we should also consider HNOO(T) => products (e.g. O(T) + HNO(S), etc.). I suppose if we used spin statistics factors, these reactions may be so slow they don't actually matter. Maybe this would clear up some of the undesired net reactions like HNO(T) + HNO(T) => HNO(S) + HNO(T).

The Schulten paper also discusses doublet + triplet, which turns up a fair amount. A fraction of 2/6 of pairs in the doublet state quickly populate the ground state (singlet + doublet). The the remaining fraction 4/6 quartet state pairs would need to "acquire doublet character by means of magnetic interactions before" populating the ground state.

rwest commented:
@donerancl @alongd  In a halogens meeting we started discussing spin-forbidden reactions (specifically an idea to add database tests to ensure that libraries don't accidentally have spin-forbidden reactions in), and my searching led me to this pull request. Was it right to mark it as stale, abandoned, and close it? Or is it a good idea worth resurrecting?

JacksonBurns commented:
@rwest, @donerancl has left the group bu @akshatzalte appears to have picked this up here: https://github.com/ReactionMechanismGenerator/RMG-Py/pull/2837 (?)
